### PR TITLE
Add PreferDefaultName property

### PR DESCRIPTION
**Pull request by maridematte**
### Problem
This is for [Issue 2303](https://github.com/dotnet/templating/issues/2303). 

### Solution
Adding support for `PreferDefaultName` optional property in `template.json`. The property defines which name should be used as a fallback option when the `-name` option is not passed in template creation.
 - If `true` the engine will use the `DefaultName` of the template.
 - If `false` the engine will use the folder's name.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=5744) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=5744) before we can accept your contribution.<br/><br/>:x: maridematte [sign now](https://cla.dotnetfoundation.org/dotnet/templating?pullRequest=5744)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/templating?pullRequest=5744) it.</sub>

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs [hunk 1]
```diff
@@ -103,7 +103,9 @@ public TemplateCreator(IEngineEnvironmentSettings environmentSettings)
                 return new TemplateCreationResult(CreationResultStatus.NotFound, templateInfo.Name, LocalizableStrings.TemplateCreator_TemplateCreationResult_Error_CouldNotLoadTemplate);
             }
 
-            string? realName = name ?? fallbackName ?? template.DefaultName;
+            string? realName = name is not null ? name : templateInfo.PreferDefaultName ? templateInfo.DefaultName : fallbackName;
+
+            // TODO: add template validation here
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs:108]: Is this still valid?

**Review by JanKrivanek — CHANGES_REQUESTED** [review]
Looks good!

Tests will be needed - you can add them to `TemplateCreatorTests`. (after https://github.com/dotnet/templating/pull/5677 is merged, it'll be possible to add snapshot tests as well - but for now best option is TemplateCreatorTests or IDE.IntegrationTests)

File: src/Microsoft.TemplateSearch.Common/TemplateDiscoveryMetadata/BlobStorageTemplateInfo.cs [hunk 2]
```diff
@@ -121,6 +121,9 @@ private BlobStorageTemplateInfo(string identity, string name, IEnumerable<string
         [JsonProperty]
         public IReadOnlyList<string> ShortNameList { get; private set; }
 
+        [JsonProperty]
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateSearch.Common/TemplateDiscoveryMetadata/BlobStorageTemplateInfo.cs:124]: We don't need it in search functionality, so it should not be written to JSON.

File: test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/Serialization/TemplateConfigModelJsonConverter.cs [hunk 3]
```diff
@@ -99,6 +96,11 @@ public override void WriteJson(JsonWriter writer, TemplateConfigModel? value, Js
                 writer.WritePropertyName("preferNameDirectory");
                 writer.WriteValue(value.PreferNameDirectory);
             }
+            if (value.PreferDefaultName)
+            {
+                writer.WritePropertyName("prefer");
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Orchestrator.RunnableProjects.UnitTests/Serialization/TemplateConfigModelJsonConverter.cs:101]:
```suggestion
                writer.WritePropertyName("preferDefaultName");
```

File: tools/Microsoft.TemplateSearch.TemplateDiscovery/Results/LegacyBlobTemplateInfo.cs [hunk 4]
```diff
@@ -102,6 +102,9 @@ public LegacyBlobTemplateInfo(ITemplateInfo templateInfo)
         [JsonProperty]
         public string Name { get; private set; }
 
+        [JsonProperty]
```
Comment (vlada-shubina) [inline tools/Microsoft.TemplateSearch.TemplateDiscovery/Results/LegacyBlobTemplateInfo.cs:105]: Same, not needed in jsons for search.

**Review by vlada-shubina — COMMENTED** [review]
Looks very good, great work! Left couple of comments inline.

The following bits are missing:
- add new field to JSON schema: https://github.com/dotnet/templating/blob/main/src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Schemas/JSON/template.json (and preferably test)
- updating documentation: https://github.com/dotnet/templating/blob/main/docs/Reference-for-template.json.md#output-management
- tests (unit and integration)

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs (hunk 1 above)
Comment (maridematte) [reply src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs]: Which part do you mean? 
 - The TODO was there to remind me to do template name validation, but I decided it was out of scope for the current PR if we wanted to change how naming works.
 - If it is the ```if else``` statement, it is to add a new logic so we do not rely first on the fallback name, and use that only as a last option.

File: docs/Reference-for-template.json.md [hunk 5]
```diff
@@ -492,11 +492,12 @@ See [the article](Binding-and-project-context-evaluation.md#bind-symbols).
...
 |`preferNameDirectory`| Boolean value, indicates whether to create a directory for the template if name is specified but an output directory is not set (instead of creating the content directly in the current directory).|If not specified, `false` is used.|no|
 |`placeholderFilename`|A filename that will be completely ignored, used to indicate that its containing directory should be copied. This allows creation of empty directory in the created template, by having a corresponding source directory containing just the placeholder file. By default, empty directories are ignored.|If not specified, a default value of `"-.-"` is used.|no|
 |`primaryOutputs`|A list of template files for further processing by the host (including post-actions). The path should contain the relative path to the file prior to the symbol based renaming that may happen during template generation. It is defined as an array of [Primary Outputs](#primary-output-definition)||no|
 
-#### Primary Output Definition
+### Primary Output Definition
```
Comment (vlada-shubina) [inline docs/Reference-for-template.json.md:500]:
```suggestion
#### Primary Output Definition
```

File: docs/Reference-for-template.json.md [hunk 6]
```diff
@@ -492,11 +492,12 @@ See [the article](Binding-and-project-context-evaluation.md#bind-symbols).
 |Name|Description|Default|Mandatory|
 |---|---|---|---|
 |`sourceName`| The text in the source content to replace with the name the user specifies. The value to be replaced with can be given using the `-n` `--name` options while running a template. The template engine will look for any occurrence of the `sourceName` present in the config file and replace it in file names and file contents. If no name is specified by the host, the current directory is used. The value of the `sourceName` is available in built-in `name` symbol and can be used as the source for creating other symbols and condition expressions. Due to implicit forms applied to `sourceName` it is important to select the replacement that doesn't cause the conflicts. See more details about `sourceName` in [the article](Naming-and-default-value-forms.md)||no|
+|`preferDefaultName`| Boolean value that decides which name will be used if no `--name` is specified during creation. If `false` it will use the fallback name, if `true` it will use the template's `defaultName`. If no `defaultName` is specified, the template creation will not succeed. | If not specified, `false` is used. | no
```
Comment (vlada-shubina) [inline docs/Reference-for-template.json.md:495]:
```suggestion
|`preferDefaultName`| Boolean value that decides which name will be used if no `--name` is specified during creation. If `false` it will use the fallback name (output folder name), if `true` it will use the template's `defaultName`. If no `defaultName` is specified and `preferDefaultName` is set to `true`, the template creation will not succeed. | If not specified, `false` is used. | no
```

what is the error in this case: no `defaultName` is specified and `preferDefaultName` is set to `true`?

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Schemas/JSON/template.json [hunk 7]
```diff
@@ -739,6 +739,10 @@
       "pattern": "^([0-9]+)$",
       "default": 0
     },
+    "preferDefaultName": {
+      "description": "Indicated which behaviour to use when no `--name` is specified during template creation.",
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Schemas/JSON/template.json:743]:
```suggestion
      "description": "Indicated which behaviour to use when no `--name` is specified during template creation. If `false` it will use the fallback (output folder) name, if `true` it will use the template's `defaultName`.",
```

File: test/Microsoft.TemplateEngine.TestTemplates/Microsoft.TemplateEngine.TestTemplates.csproj [hunk 8]
```diff
@@ -21,4 +21,8 @@
     <Content Include="nupkg_templates\**" />
     <None Include="nupkg_templates\**" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="test_templates\TemplateWithPreferDefaultNameButNoDefaultName\.template.config\template.json" />
+    <None Remove="test_templates\TemplateWithPreferDefaultNameButNoDefaultName\toChange.cs" />
+  </ItemGroup>
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.TestTemplates/Microsoft.TemplateEngine.TestTemplates.csproj:27]:
```suggestion
```

File: src/Microsoft.TemplateEngine.Edge/Settings/TemplateInfoReader.cs [hunk 9]
```diff
@@ -41,6 +41,7 @@ internal static TemplateInfo FromJObject(JObject entry)
                 }
 
                 info.DefaultName = entry.ToString(nameof(DefaultName));
+                info.PreferDefaultName = entry.ToBool(nameof(PreferDefaultName));
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/Settings/TemplateInfoReader.cs:44]: consider the tests for template cache read/write.
See https://github.com/dotnet/templating/blob/main/test/Microsoft.TemplateEngine.Edge.UnitTests/TemplateCacheTests.cs

File: docs/Reference-for-template.json.md (hunk 6 above)
Comment (maridematte) [reply docs/Reference-for-template.json.md:495]: It will result in a `MissingMandatoryParam` status if there is no fallback name specified. If there is a fallback name, then it will be that name. I realize that this is not clear in the docs, so I will adjust it accordingly.

**Review by JanKrivanek — APPROVED** [review]
Looks solid to me - good job!

I added few comments o consider for making test code more conscise.
You can probably as well use templates test FW (via passing explicit null as name parameter).
After this is merged please create PR in schemastore to reflect your changes there as well - https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/template.json

Overall it's great!!

File: docs/Reference-for-template.json.md (hunk 6 above)
Comment (vlada-shubina) [reply docs/Reference-for-template.json.md]: It would be better if it results in `TemplateIssueDetected`  status https://github.com/dotnet/templating/blob/c0f7124770e08c67b86197cb655c7eed7410096c/src/Microsoft.TemplateEngine.Edge/Template/CreationResultStatus.cs#L26 
as it is an authoring error with correct error message. The user cannot fix it - it's authoring error. 

@JanKrivanek please correct me if the status i mentioned  is wrong.
We should also added the check for this in https://github.com/dotnet/templating/issues/2623

Comment (JanKrivanek) [reply docs/Reference-for-template.json.md]: That's correct - `TemplateIssueDetected` would be better fit in this case.

Btw. it seems we are missing the `defaultName` in this doc. @maridematte - can you please add the `defaultName` into the `Template Definition` section (around the line 40)? It might be as well helpful to crerate html anchor there and link it from `defaultName` mention on this line.

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs [hunk 10]
```diff
@@ -103,10 +103,12 @@ public TemplateCreator(IEngineEnvironmentSettings environmentSettings)
                 return new TemplateCreationResult(CreationResultStatus.NotFound, templateInfo.Name, LocalizableStrings.TemplateCreator_TemplateCreationResult_Error_CouldNotLoadTemplate);
             }
 
-            string? realName = name ?? fallbackName ?? template.DefaultName;
+            string? realName = name is not null ? name : templateInfo.PreferDefaultName ? templateInfo.DefaultName : null;
+            realName = realName is null ? fallbackName : realName;
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs:107]: Second look on this logic - it doesn't look correct.
CLI always passes the name of current directory as the fallback name. 
Therefore according to this logic:
- if name is `null`
- and if `PreferDefaultName` is true
- and if `templateInfo.DefaultName` is null

`fallbackName` will be still used, whereas it shouldn't.

if looks like it should be changed as following
```
var finalName = null;

if (!string.IsNullOrEmpty(name))
{
    //name was specified via "name" parameter - use it
    finalName = name;
}
//use default name
else if (templateInfo.PreferDefaultName)
{
   if (string.IsNullOrEmpty(templateInfo.DefaultName)
   {
      return new TemplateCreationResult(CreationResultStatus.TemplateIssueDetected, template.Name, "The template authoring issue: the template has `PreferDefaultName` set to `true`, but `DefaultName` was not configured.")
   }
   finalName = templateInfo.DefaultName;
}
//use fallback name if available
else
{
   if (string.IsNullOrEmpty(fallbackName)
   {
        if (string.IsNullOrEmpty(templateInfo.DefaultName))
       {
           //error - nothing to use as the name
        }
        else
        {
            finalName = templateInfo.DefaultName;
        }
   }
   else
   {
          finalName = fallbackName;
   }
}
```
However, on the other hand `name` is not always used, so i have doubts if it should be mandatory. But since current logic implies it is mandatory, let's keep it as. 
CLI has only few cases where `fallbackName` will be null. We should also check what VS can send as `fallbackName`.

File: src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt [hunk 11]
```diff
@@ -1,5 +1,9 @@
...
 Microsoft.TemplateEngine.Abstractions.IVariableCollection.Parent.get -> Microsoft.TemplateEngine.Abstractions.IVariableCollection?
 Microsoft.TemplateEngine.Abstractions.IVariableCollection.Parent.set -> void
-Microsoft.TemplateEngine.Abstractions.ITemplateInfo.PreferDefaultName.get -> bool
\ No newline at end of file
+Microsoft.TemplateEngine.Abstractions.ITemplateInfo.PreferDefaultName.get -> bool
+Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions
+static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetData.Empty.get -> Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData!
+static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions.GetValue(this Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData! data, string! parameterName) -> Microsoft.TemplateEngine.Abstractions.Parameters.ParameterData!
+static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions.TryGetValue(this Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData! data, string! parameterName, out Microsoft.TemplateEngine.Abstractions.Parameters.ParameterData? parameterData) -> bool
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt:9]: are those changes needed? seems irrelevant to the change

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs [hunk 12]
```diff
@@ -105,10 +105,36 @@ public TemplateCreator(IEngineEnvironmentSettings environmentSettings)
...
             realName = realName is null ? fallbackName : realName;
+            if (!string.IsNullOrEmpty(name))
+            {
+                realName = name;
+            }
+            else if (templateInfo.PreferDefaultName)
+            {
+                if (string.IsNullOrEmpty(templateInfo.DefaultName))
+                {
+                    return new TemplateCreationResult(CreationResultStatus.TemplateIssueDetected, template.Name, "--name");
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs:116]: `TemplateCreationResult` should contain the localized error message specifying the issue and call to action to resolve it.

In case of `dotnet new` this message will be printed out here: https://github.com/dotnet/sdk/blob/0fdccd38a185af6d8f9fa74d7d386ebdef2c58d7/src/Cli/Microsoft.TemplateEngine.Cli/TemplateInvoker.cs#L283-L288

Now it looks as only `--name` will be printed.

Could you please check this?

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs [hunk 13]
```diff
@@ -103,7 +103,35 @@ public TemplateCreator(IEngineEnvironmentSettings environmentSettings)
                 return new TemplateCreationResult(CreationResultStatus.NotFound, templateInfo.Name, LocalizableStrings.TemplateCreator_TemplateCreationResult_Error_CouldNotLoadTemplate);
             }
 
-            string? realName = name ?? fallbackName ?? template.DefaultName;
+            string? realName = name is not null ? name : templateInfo.PreferDefaultName ? templateInfo.DefaultName : null;
+            realName = realName is null ? fallbackName : realName;
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs:107]: are those lines needed? `realName` will be reassigned in the following code.

File: src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt (hunk 11 above)
Comment (maridematte) [reply src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt:9]: The one on about the `PreferDefaultName` was added automatically to the file by the IDE. So I'm assuming that is a standard for the code. The rest came from the main branch, so I'm not sure what they are about (and not sure why they are showing separately)

File: src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs (hunk 13 above)
Comment (maridematte) [reply src/Microsoft.TemplateEngine.Edge/Template/TemplateCreator.cs:107]: oops, forgot to change that after, thanks for catching it!

File: src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx [hunk 14]
```diff
@@ -351,6 +351,10 @@ Details: {1}.</value>
     <value>Failed to create template.
 Details: {0}</value>
   </data>
+  <data name="TemplateCreator_TemplateCreationResult_Error_NoDefaultName" xml:space="preserve">
+    <value>Failed to create template due naming issues</value>
+    <comment> Template does not contain a DefaultName that can be used when --name is not passed </comment>
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx:356]:
```suggestion
   <comment>Template configuration requires the name to specified but does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating the template.</comment>
```

Please avoid `--name` syntax, as it is only applicable for CLI. Just "name" should be good enough.

File: src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt [hunk 15]
```diff
@@ -1,8 +1,9 @@
-﻿Microsoft.TemplateEngine.Abstractions.ITemplate.TemplateSourceRoot.get -> Microsoft.TemplateEngine.Abstractions.Mount.IDirectory!
+Microsoft.TemplateEngine.Abstractions.ITemplate.TemplateSourceRoot.get -> Microsoft.TemplateEngine.Abstractions.Mount.IDirectory!
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt:1]: consider removing white-space only change

File: src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt [hunk 16]
```diff
@@ -1,8 +1,9 @@
...
+Microsoft.TemplateEngine.Abstractions.ITemplateInfo.PreferDefaultName.get -> bool
 Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions
 static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetData.Empty.get -> Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData!
 static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions.GetValue(this Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData! data, string! parameterName) -> Microsoft.TemplateEngine.Abstractions.Parameters.ParameterData!
-static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions.TryGetValue(this Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData! data, string! parameterName, out Microsoft.TemplateEngine.Abstractions.Parameters.ParameterData? parameterData) -> bool
\ No newline at end of file
+static Microsoft.TemplateEngine.Abstractions.Parameters.ParameterSetDataExtensions.TryGetValue(this Microsoft.TemplateEngine.Abstractions.Parameters.IParameterSetData! data, string! parameterName, out Microsoft.TemplateEngine.Abstractions.Parameters.ParameterData? parameterData) -> bool
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Abstractions/PublicAPI.Unshipped.txt:9]: consider removing white-space only change

File: test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs [hunk 17]
```diff
@@ -367,5 +367,61 @@ common text
...
+            Dictionary<string, string?> parameters = new()
+            {
+                { "some", "parameter" },
+            };
+
+            ITemplateCreationResult result = await bootstrapper
+                .CreateAsync(foundTemplates[0].Info, null, output, parameters)
+                .ConfigureAwait(false);
+
+            Assert.Equal(CreationResultStatus.TemplateIssueDetected, result.Status);
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs:424]: nit: it would be good to check error message as well

File: test/Microsoft.TemplateEngine.Edge.UnitTests/TemplateCreatorTests.cs [hunk 18]
```diff
@@ -670,12 +671,68 @@ public async void InstantiateAsync_ConditionalParametersIsRequiredEvaluation(boo
...
+                "preferDefaultName": false,
+                "sourceName": "sourceFile"
+            }
+            """;
+
+        [Theory]
+        [InlineData(TemplateConfigPreferDefaultNameWithDefaultName, "thisIsAName", "./thisIsAName.cs", false, "")]
+        [InlineData(TemplateConfigPreferDefaultNameWithDefaultName, null, "./defaultName.cs", false, "")]
+        [InlineData(TemplateConfigNoPreferDefaultNameWithDefaultName, null, "./tst2.cs", false, "")]
+        [InlineData(TemplateConfigPreferDefaultNameWithoutDefaultName, null, "./tst2.cs", true, "Failed to create template due naming issues")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Edge.UnitTests/TemplateCreatorTests.cs:709]: Is this error message correct? Where it's coming from? Should it be the one from `LocalizableStrings` now?

File: src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx [hunk 19]
```diff
@@ -351,6 +351,10 @@ Details: {1}.</value>
     <value>Failed to create template.
 Details: {0}</value>
   </data>
+  <data name="TemplateCreator_TemplateCreationResult_Error_NoDefaultName" xml:space="preserve">
+    <value>Failed to create template due naming issues</value>
+    <comment>Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating</comment>
+  </data>
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx:357]:
```suggestion
  <data name="TemplateCreator_TemplateCreationResult_Error_NoDefaultName" xml:space="preserve">
    <value>Failed to create template: the template configuration does not define a default name. Specify the name for the template when instantiating or define default name in the template configuration.</value>
    <comment></comment>
  </data>
```

The comment field is for translator only.

File: src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx (hunk 14 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx]: Ignore this suggestion please - my fault, the error message should be in `value` field. `comment` is just for translator. See https://github.com/dotnet/templating/pull/5744#discussion_r1058854329 instead.

File: src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx [hunk 20]
```diff
@@ -351,6 +351,9 @@ Details: {1}.</value>
     <value>Failed to create template.
 Details: {0}</value>
   </data>
+  <data name="TemplateCreator_TemplateCreationResult_Error_NoDefaultName" xml:space="preserve">
+    <value>Failed to create template: the Template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.</value>
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Edge/LocalizableStrings.resx:355]:
```suggestion
    <value>Failed to create template: the template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.</value>
```

File: test/Microsoft.TemplateEngine.Edge.UnitTests/TemplateCreatorTests.cs [hunk 21]
```diff
@@ -670,12 +671,68 @@ public async void InstantiateAsync_ConditionalParametersIsRequiredEvaluation(boo
...
+                "preferDefaultName": false,
+                "sourceName": "sourceFile"
+            }
+            """;
+
+        [Theory]
+        [InlineData(TemplateConfigPreferDefaultNameWithDefaultName, "thisIsAName", "./thisIsAName.cs", false, "")]
+        [InlineData(TemplateConfigPreferDefaultNameWithDefaultName, null, "./defaultName.cs", false, "")]
+        [InlineData(TemplateConfigNoPreferDefaultNameWithDefaultName, null, "./tst2.cs", false, "")]
+        [InlineData(TemplateConfigPreferDefaultNameWithoutDefaultName, null, "./tst2.cs", true, "Failed to create template: the Template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.")]
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.Edge.UnitTests/TemplateCreatorTests.cs:709]:
```suggestion
        [InlineData(TemplateConfigPreferDefaultNameWithoutDefaultName, null, "./tst2.cs", true, "Failed to create template: the template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.")]
```

File: test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs [hunk 22]
```diff
@@ -367,5 +367,64 @@ common text
...
+            Assert.Equal(
+                "Failed to create template: the Template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.",
```
Comment (vlada-shubina) [inline test/Microsoft.TemplateEngine.IDE.IntegrationTests/End2EndTests.cs:426]:
```suggestion
                "Failed to create template: the template name is not specified. Template configuration does not configure a default name that can be used when name is not specified. Specify the name for the template when instantiating or configure a default name in the template configuration.",
```

**Review by vlada-shubina — APPROVED** [review]
Thanks, looks good to go.
Just a small casing issue in the error message
